index.js

```diff
@@ -1,201 +1,114 @@
-            }
-        }
-    );
-}
 
-function checkIfInitialized() {
-    return RNOneSignal != null;
-}
+    /* O B S E R V E R S */
 
-export default class OneSignal {
+    static addPermissionObserver(observer) {
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
 
-    /**
-     Listen to events of received, opened, ids, subscription, permission, emailSubscription, inAppMessageClicked
-     TODO: We currently have implemented the steps up until connecting the "SUBSCRIPTION_EVENT" and "PERMISSION_EVENT"
-     Currently the getPermissionSubscriptionState is used to get all device information and
-        needs to be broken up into using the native observers to fire these React-Native handlers
-     */
-    static addEventListener(type, handler) {
-        if (!checkIfInitialized()) return;
-
-        invariant(
-            type === NOTIFICATION_RECEIVED_EVENT ||
-            type === NOTIFICATION_OPENED_EVENT ||
-            type === IDS_AVAILABLE_EVENT ||
-//            type === SUBSCRIPTION_EVENT ||
-//            type === PERMISSION_EVENT ||
-            type === EMAIL_SUBSCRIPTION_EVENT ||
-            type === IN_APP_MESSAGE_CLICKED_EVENT,
-            'OneSignal only supports received, opened, ids, emailSubscription, and inAppMessageClicked events'
-        );
-
-        _eventTypeHandler.set(type, handler);
-
-        // Make native request to init notification opened handler
-        if (type === NOTIFICATION_OPENED_EVENT) {
-            RNOneSignal.initNotificationOpenedHandlerParams();
+        if (Platform.OS === 'android') {
+            RNOneSignal.addPermissionObserver();
         }
+        eventManager.addEventHandler(PERMISSION_CHANGED, observer);
+    }
 
-        // Make native request to init idsAvailable handler
-        if (type === IDS_AVAILABLE_EVENT) {
-            RNOneSignal.idsAvailable();
-        }
+    static addSubscriptionObserver(observer) {
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
 
-        // Make native request to init IAM handler
-        if (type === IN_APP_MESSAGE_CLICKED_EVENT) {
-            if (Platform.OS === 'android') {
-                RNOneSignal.initInAppMessageClickHandlerParams();
-            } else if (Platform.OS === 'ios') {
-                RNOneSignal.setInAppMessageClickHandler();
-            }
+        if (Platform.OS === 'android') {
+            RNOneSignal.addSubscriptionObserver();
         }
+        eventManager.addEventHandler(SUBSCRIPTION_CHANGED, observer);
+    }
 
-        // Check if there is a cache for this type of event
-        var cache = _notificationCache.get(type);
-        if (handler && cache) {
-            handler(cache);
-            _notificationCache.delete(type);
+    static addEmailSubscriptionObserver(observer) {
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
+
+        if (Platform.OS === 'android') {
```

Comment:
Remove android check, this method is available for iOS as well.

Reply:
Done